services/ssh.nix

```diff
@@ -1,15 +1,49 @@
-{
-  services.openssh.enable = true;
+{ pkgs, ... }:
+let
+  whodir = "/run/whoroot";
+
+  # A wrapper over w which replaces root with the username from the key
+  # of the logged in rooter. The files in whodir are written by the
+  # loginShellInit
+  # The first regex removes the slash from the tty
+  whoroot = pkgs.writeShellScriptBin "wroot" ''
+    sedscript="s!pts/!pts!g;"
+    for f in ${whodir}/*; do
+      bf=$(basename $f)
+      sedscript="''${sedscript}s!root +$bf!$(cat $f)\t$bf!g;"
+    done
+    ${pkgs.procps}/bin/w | ${pkgs.gnused}/bin/sed -E "$sedscript"
+  '';
+in {
+  services.openssh = {
+    enable = true;
+    extraConfig = ''
+      PermitUserEnvironment yes
+    '';
+  };
+
   users.users.root.openssh.authorizedKeys.keys = [
-    "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAINnVLSh0OStxZTkXE6oGgwfFvsbvN6bFPlVfDYOwtnzn m1cr0man@@redbrick.dcu.ie"
-    "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIPR+OTAIYr02f/WKQSXo7zYy9tkuAHYpy0ajqY6aJ7Nk m1cr0man@redbrick.dcu.ie"
-    "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIDYKnYP4Mmyk4wQE7J6Tyr27XToKtxAhXBZr5HkEXiFq m1cr0man@redbrick.dcu.ie"
-    "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIHvabMrrJILDua2sedVqBStb6YKBHpgCO5HOM98l7uwf greenday@redbrick.dcu.ie"
-    "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIOjro8OS7cWf6xBcrs4erZqjN5JdztoGqpMXFQwzd9pV mctastic@azazel.redbrick.dcu.ie"
-    "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDLSLIF2IOo/OzbmbMGp4kt6VP2z8zNCuuVNyuxyBU0A8cOeUhkAbVibVmFqPlcHDJ4+zhkNN0GDnEJEUAmBNi+yc9EJG7StxdguEAKPlA9gQ/Z73cMrfMHtTPOHj/uCKUqi9vzb3tlOltJmuS3SwF0B5dk58j/cwr3nEEzikMmQIykxI+F+rxMnxaQXtNBGz3ednAaJ4Lvv9JSxWcExEtU0lM0X1MgZgkYFr48uQwsDUE+j23+wifMrOA+zhC0uRcuIapnxsyoW/wDOYrQZFlw6acrVX+zNtxcCQoqIX4oAobCXn7tYz9peKrV8TmJwQOspsmyY75xIAbyz0AiD3oh ylmcc@redbrick.dcu.ie"
-    "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIG/8Xf5/DtcOPjZKfag4ATBe5a3I1HvhYqi8fV7si4OU butlerx@redbrick.dcu.ie"
-    "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIIhA5mm1sBzz6tcrUF2FzW6wrckW1IsQAyS8Bfu4yJRJ d_fens@redbrick.dcu.ie"
-    "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIGybjW48+tQaykqDIuSeuH/3GLQRHZDa1toJOIB/FrD4 fraz@redbrick.dcu.ie"
-    "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIKVc8zqnZkzsOHzfycpJ3QbB9SJ2FxmRRifYbBuuixk2 galvinio@redbrick.dcu.ie"
+    "environment=\"REMOTEUSER=m1cr0man\" ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAINnVLSh0OStxZTkXE6oGgwfFvsbvN6bFPlVfDYOwtnzn m1cr0man@redbrick.dcu.ie"
```

Comment:
Could we use the ssh key email?

Reply:
Do you mean add it to REMOTEUSER? Eh, it'll make the wroot output a bit long.

Reply:
i mean parse the email field ranter then annotae every line with `REMOTEUSER` seems redundent to have both

Reply:
No. I'm not parsing this field in the first place, OpenSSH is setting env vars based on the key used. I can't tell from the shell what key was used otherwise so I wouldn't be able to grep the line to get the email field based on the public key.